KAFKA-10816: Add health check endpoint for Kafka Connect

C0urante opened the pull request:
[Jira](https://issues.apache.org/jira/browse/KAFKA-10816), [KIP-1017](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1017%3A+Health+check+endpoint+for+Kafka+Connect)

Adds the `GET /health` endpoint to the Kafka Connect REST API, and covers all three states (starting, healthy, unhealthy) in both modes (distributed, standalone) with integration tests. Some unit tests are also added.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
KAFKA-10816: Add health check endpoint for Kafka Connect

C0urante commented:
@gharris1727, if you have time would you mind taking a look? I know the diff is fairly large but most of that is either small-but-far-reaching changes (i.e., getting rid of the `DistributedHerder::running` field and replacing its uses in our integration tests with the health check endpoint) or newly-added unit and integration tests. Still, if this is too much to do in one round, please let me know and I can try to break it out into smaller PRs.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public static WorkerStatus healthy() {
+        return new WorkerStatus(
```
nit: have a single instance that is always returned? I like the method, rather than having a final constant.

C0urante replied:
👍 done

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/WorkerStatus.java:
```diff
@@ -0,0 +1,99 @@
+        );
+    }
+
+    public static WorkerStatus starting(Stage stage) {
```
These are pretty natural methods, and I like them. And even though these are not part of the public api...

I think it is a bit odd that these rest entities would be aware of an internal abstraction `Stage`. It's not the first time, as TaskInfo depends on ConnectorTaskId, ServerInfo on AppInfoParser, etc.

Just a thought that crossed my mind, but I don't have an alternative that I like better so this can probably stay as-is.

C0urante replied:
It should be easy enough to change the parameter to something like `String statusDetail` and leave it to the caller to invoke `Stage::summarize` if they're using the stage class to track progress. Did that not seem better to you? (I'd be fine with it.)

C0urante replied:
I gave this a shot, let me know what you think.

gharris1727 replied:
Ah yeah this is very reasonable. For some reason I thought there were a lot of call-sites that would need an extra null guard, but that wasn't the case. This is great.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -124,10 +127,21 @@ public synchronized void stop() {
         }
         stopServices();
-        running = false;
+        healthCheckThread.shutDown();
```
Please call `Thread#join()` at some point to ensure the thread gets cleaned up :)

C0urante replied:
Ah, good call. Done 👍

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/RootResource.java:
```diff
@@ -32,15 +43,59 @@
+                workerStatus = WorkerStatus.unhealthy(stage);
+            }
+        } catch (ExecutionException e) {
+            throw e.getCause();
```
Should this be wrapped in a ConnectRestException, so it gets picked up by the ConnectExceptionMapper? It falls back to a 500 otherwise so it doesn't have a lot of difference, and it should never throw an ExecutionException, but 🤷

C0urante replied:
The exception mapper will catch this anyways, won't it? The intent was to hit [this part](https://github.com/apache/kafka/blob/d0dfefbe6394276eb329b6ca998842a984add506/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/errors/ConnectExceptionMapper.java#L73-L77) of the code path and generate the standard 500-on-unexpected-error response.

gharris1727 commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnect.java:
```diff
@@ -129,14 +129,17 @@ public void start() {
             throw new ConnectException("Failed to start HTTP client", e);
         }
 
+        startConnect();
```
Does the order here matter? An exception starting httpClient still leaves the kafkaCluster started.

C0urante replied:
We need to start the HTTP client before we start the Connect cluster in order to be able to hit the cluster's REST API before `startConnect` has returned (which is only possible in the standalone integration test because we run `EmbeddedConnectStandalone::start` on a separate thread from the one we use to make REST requests).

This is dirty-bordering-on-filthy, but I couldn't think of a better way to probe a worker's health check endpoint during simulated blocked startup. Thoughts?

gharris1727 commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -226,25 +240,37 @@ private void stop(boolean deleteLogDirs, boolean stopZK) {
-                    log.error(msg, t);
-                    throw new RuntimeException(msg, t);
-                }
+    private void awaitBrokerShutdown() {
```
Was there a problem where the log directories would be deleted before the brokers shut down? Nice find!

C0urante replied:
Nah, this was just because there was a race in one of our integration tests where we shut down the underlying Kafka cluster and then wait for a fixed time period before asserting that the Connect cluster is unavailable. Making Kafka cluster shutdown synchronous fixed that problem.

(And the method decomposition was just to appease Checkstyle, although I do have to admit I also find it easier to read now.)

C0urante replied:
I gave this a shot, let me know what you think.

gharris1727 commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/WorkerHandle.java:
```diff
@@ -54,24 +55,6 @@ public void stop() {
-     *
-     * @return the worker's name
-     */
-    public String name() {
```
This appears to have been unused. I can't imagine a situation where you'd want to programmatically search for the name, so it should probably only be visible in the toString() :+1:

gharris1727 reviewed (commented):
Thanks for the feature, Chris!

It mostly looks good to me. I had a few minor comments i'd like your thoughts on.

C0urante committed:
Introduce WorkerStatus.HEALTHY constant

C0urante committed:
Extract HealthCheckThread to separate class, add call to Thread::join in HealthCheckThread::shutDown

C0urante committed:
Change static factory constructors for WorkerStatus to accept String instead of Stage

C0urante commented:
Thanks for the review @gharris1727! This is ready for another pass when you have time.

gharris1727 reviewed (approved):
LGTM, thanks @C0urante!

C0urante committed:
Remove unnecessary TODO

There is no need to increase the REST request timeout, since we're asserting
that the worker is unavailable when we issue the request anyways. There is also
no risk that the REST server has not been started or initialized yet, since both
of those operations take place synchronously during
EmbeddedConnectCluster::addWorker.

C0urante commented:
Thanks Greg!

I've realized the TODO in the code base was unnecessary and this should be safe to merge without waiting for more green CI runs (see the latest commit message for more details).

Merging...
